crates/common/mqtt_channel/src/connection.rs

```diff
@@ -287,11 +301,26 @@ impl Connection {
                     }
                 }
 
-                Ok(Event::Incoming(Incoming::Disconnect))
+                Ok(Event::Incoming(Packet::Disconnect))
                 | Ok(Event::Outgoing(Outgoing::Disconnect)) => {
                     break;
                 }
 
+                Ok(Event::Outgoing(Outgoing::Publish(p))) => {
+                    if !awaiting_ack.contains(&p) {
+                        pub_count.fetch_sub(1, Ordering::SeqCst);
+                    }
+                    awaiting_ack.insert(p);
+                }
+
+                Ok(Event::Incoming(Packet::PubAck(p))) => {
+                    awaiting_ack.remove(&p.pkid);
+                }
+
+                Ok(Event::Incoming(Packet::PubRec(p))) => {
```

Comment:
Isn't it safer to remove the packet once `PUBCOMP` is received rather than PUBREC? Since that guarantees that the broker has released the message to other subscribers?
```suggestion
                Ok(Event::Incoming(Packet::PubComp(p))) => {
```

Reply:
Yes this is probably true

Reply:
Actually true, as `PUBCOMP` is the QoS 2 equivalent of `PUBACK` of QoS 1. By closing on `PUBREC`, the `PUBREL` might be missing and the broker let waiting for it.